### ISSUE TITLE
Trin db strategy

### ISSUE THREAD

**Issue by njgheorghita** [ethereum/trin#91]
This is just an issue to track an ongoing design decision regarding trin and its db strategy.
tl;dr: Is there one central db that subnets share, or is each subnet responsible for its own db.

Some thoughts so far...
@jacobkaufmann 
> As far as the database stuff goes, my assumption is that we would require multiple database instances because we would need to manage the storage capacity for each subnetwork individually. For example, a user may wish to allocate X space to state and Y space to history. I'm not sure this is possible with a single database instance, but I haven't looked into the RocksDB crate enough to be absolutely sure.

@mrferris 
> On the topic of storage, my thoughts are that we have separate database instances per network, each abstracted by a separate PortalStorage instance (#69) per network.
My main reasoning is simplicity. If we have every network's data combined in one DB instance, we'd need to store metadata on each key to denote which network it belongs to. The PortalStorage struct would need to get more complex in order to keep track of everything for every network at the same time....or we could have multiple PortalStorage instances interacting with a single database instance, in which case we need to manage locking etc...debugging storage functionality also gets more complex with everything in one RocksDB.
Is there a benefit to combining the databases that I might be missing? I'm only seeing the extra complexity. That said, I have yet to look into the effects of multiple DBs vs 1 DB on disk i/o, so that might be what I'm missing.

@carver
> I think most users won't want to manage the storage space of each subnetwork individually (they probably won't even know the internal detail that there are multiple subnetworks). Think of a typical bittorrent user, and how little they probably know about the details of the protocol.
Even if we decide to implement this per-network limit, I'm sure we could find other ways than running multiple DBs. So I think we should at least remain open to the idea of piping everything into a single DB (which I expect to have performance benefits due to I/O serialization).
Yup, that's exactly the issue. Some client (Erigon?) had to do a big rewrite to a single DB in order to avoid causing big I/O delays by trying to hammer the disk a bunch in parallel from different threads.
FWIW, I think we're going to have to deal with locking logic anyway, even within a single subnet.

**Comment by njgheorghita** [comment]
Closing since this discussion is being tracked in #240
